Unable to fully read images of an epub FXL

I tried to open a comic book packaged into a fixed layout epub but each image fails to be displayed.
The image is very slowly drawn but is stopped in the middle of the process.

The device: OnePlus One (Android 6.0.1 / Cyanogen OS / 3Go RAM).
The app: r2-testapp 2.0.0-alpha.2

The stack trace:
```
10-29 09:17:18.789 5578-11276/org.readium.r2reader E/NanoHTTPD: Could not send response to the client
    java.io.IOException: Can't read ResourceInputStream
        at org.readium.r2.shared.fetcher.ResourceInputStream.read(ResourceInputStream.kt:107)
        at org.nanohttpd.protocols.http.response.Response.sendBody(Response.java:336)
        at org.nanohttpd.protocols.http.response.Response.sendBodyWithCorrectEncoding(Response.java:313)
        at org.nanohttpd.protocols.http.response.Response.sendBodyWithCorrectTransferAndEncoding(Response.java:300)
        at org.nanohttpd.protocols.http.response.Response.send(Response.java:268)
        at org.nanohttpd.protocols.http.HTTPSession.execute(HTTPSession.java:432)
        at org.nanohttpd.protocols.http.ClientHandler.run(ClientHandler.java:75)
        at java.lang.Thread.run(Thread.java:818)
     Caused by: org.readium.r2.shared.fetcher.Resource$Exception$Other: java.lang.Exception: Unable to skip enough bytes
        at org.readium.r2.shared.fetcher.Resource$Exception$Companion.wrap(Resource.kt:192)
        at org.readium.r2.shared.fetcher.ArchiveFetcher$EntryResource.read(ArchiveFetcher.kt:114)
        at org.readium.r2.shared.fetcher.ArchiveFetcher$EntryResource$read$1.invokeSuspend(ArchiveFetcher.kt)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
        at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
        at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
        at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
        at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
        at org.readium.r2.shared.fetcher.ResourceInputStream.read(ResourceInputStream.kt:95)
        at org.nanohttpd.protocols.http.response.Response.sendBody(Response.java:336) 
        at org.nanohttpd.protocols.http.response.Response.sendBodyWithCorrectEncoding(Response.java:313) 
        at org.nanohttpd.protocols.http.response.Response.sendBodyWithCorrectTransferAndEncoding(Response.java:300) 
        at org.nanohttpd.protocols.http.response.Response.send(Response.java:268) 
        at org.nanohttpd.protocols.http.HTTPSession.execute(HTTPSession.java:432) 
        at org.nanohttpd.protocols.http.ClientHandler.run(ClientHandler.java:75) 
        at java.lang.Thread.run(Thread.java:818) 
     Caused by: java.lang.Exception: Unable to skip enough bytes
        at org.readium.r2.shared.extensions.InputStreamKt$readRange$2.invokeSuspend(InputStream.kt:70)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

I can provide the epub file in private.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,12 @@ All notable changes to this project will be documented in this file.
 
 ## [Unreleased]
 
+### Fixed
+
+* [#129](https://github.com/readium/r2-shared-kotlin/issues/129) Improve performances when reading deflated ZIP resources.
+  * For example, it helps with large image-based FXL EPUB which used to be slow to render.
+
+
 ## [2.0.0-beta.1]
 
 ### Added

--- a/r2-shared/src/main/java/org/readium/r2/shared/fetcher/ArchiveFetcher.kt
+++ b/r2-shared/src/main/java/org/readium/r2/shared/fetcher/ArchiveFetcher.kt
@@ -82,7 +82,11 @@ class ArchiveFetcher private constructor(private val archive: Archive) : Fetcher
             metadataLength()?.let { Try.success(it) }
                 ?: read().map { it.size.toLong() }
 
-        override suspend fun close() {}
+        override suspend fun close() {
+            if (::_entry.isInitialized) {
+                _entry.onSuccess { it.close() }
+            }
+        }
 
         private suspend fun metadataLength(): Long? =
             entry().getOrNull()?.length

--- a/r2-shared/src/main/java/org/readium/r2/shared/util/archive/Archive.kt
+++ b/r2-shared/src/main/java/org/readium/r2/shared/util/archive/Archive.kt
@@ -68,6 +68,11 @@ interface Archive {
          */
         suspend fun read(range: LongRange? = null): ByteArray
 
+        /**
+         * Closes any pending resources for this entry.
+         */
+        suspend fun close()
+
     }
 
     /** List of all the archived file entries. */

--- a/r2-shared/src/main/java/org/readium/r2/shared/util/archive/ExplodedArchive.kt
+++ b/r2-shared/src/main/java/org/readium/r2/shared/util/archive/ExplodedArchive.kt
@@ -43,6 +43,9 @@ internal class ExplodedArchive(private val directory: File)  : Archive {
                     it.readRange(range)
             }
         }
+
+        override suspend fun close() {}
+
     }
 
     override suspend fun entries(): List<Archive.Entry> =

--- a/r2-shared/src/main/java/org/readium/r2/shared/util/archive/JavaZip.kt
+++ b/r2-shared/src/main/java/org/readium/r2/shared/util/archive/JavaZip.kt
@@ -13,7 +13,9 @@ import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.withContext
 import org.readium.r2.shared.extensions.readFully
 import org.readium.r2.shared.extensions.readRange
+import org.readium.r2.shared.util.io.CountingInputStream
 import java.io.File
+import java.io.InputStream
 import java.lang.Exception
 import java.util.zip.ZipEntry
 import java.util.zip.ZipFile
@@ -32,17 +34,53 @@ internal class JavaZip(private val archive: ZipFile) : Archive {
                 else
                     entry.compressedSize.takeUnless { it == -1L }
 
-        override suspend fun read(range: LongRange?): ByteArray {
-            val stream = withContext(Dispatchers.IO) {
-                archive.getInputStream(entry)
-            }
-            return stream.use {
-                if (range == null)
-                    it.readFully()
-                else
-                    it.readRange(range)
+        override suspend fun read(range: LongRange?): ByteArray =
+            if (range == null)
+                readFully()
+            else
+                readRange(range)
+
+        private suspend fun readFully(): ByteArray = withContext(Dispatchers.IO) {
+            archive.getInputStream(entry).use {
+                it.readFully()
             }
         }
+
+        private suspend fun readRange(range: LongRange): ByteArray =
+            stream(range.first).readRange(range)
+
+        /**
+         * Reading an entry in chunks (e.g. from the HTTP server) can be really slow if the entry
+         * is deflated in the archive, because we can't jump to an arbitrary offset in a deflated
+         * stream. This means that we need to read from the start of the entry for each chunk.
+         *
+         * To alleviate this issue, we cache a stream which will be reused as long as the chunks are
+         * requested in order.
+         *
+         * See this issue for more info: https://github.com/readium/r2-shared-kotlin/issues/129
+         */
+        private suspend fun stream(fromIndex: Long): CountingInputStream {
+            // Reuse the current stream if it didn't exceed the requested index.
+            stream
+                ?.takeIf { it.count <= fromIndex }
+                ?.let { return it }
+
+            return withContext(Dispatchers.IO) {
+                val newStream = CountingInputStream(archive.getInputStream(entry))
+                stream?.close()
+                stream = newStream
+                newStream
+            }
+        }
+
+        private var stream: CountingInputStream? = null
+
+        override suspend fun close() {
+            withContext(Dispatchers.IO) {
+                stream?.close()
+            }
+        }
+
     }
 
     override suspend fun entries(): List<Archive.Entry> =

--- a/r2-shared/src/main/java/org/readium/r2/shared/util/io/CountingInputStream.kt
+++ b/r2-shared/src/main/java/org/readium/r2/shared/util/io/CountingInputStream.kt
@@ -1,0 +1,82 @@
+/*
+ * Copyright 2021 Readium Foundation. All rights reserved.
+ * Use of this source code is governed by the BSD-style license
+ * available in the top-level LICENSE file of the project.
+ */
+
+package org.readium.r2.shared.util.io
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.readium.r2.shared.extensions.coerceFirstNonNegative
+import org.readium.r2.shared.extensions.read
+import org.readium.r2.shared.extensions.requireLengthFitInt
+import java.io.FilterInputStream
+import java.io.IOException
+import java.io.InputStream
+import java.lang.Exception
+
+/**
+ * An [InputStream] counting the number of bytes read from a wrapped [inputStream].
+ */
+class CountingInputStream(inputStream: InputStream) : FilterInputStream(inputStream) {
+
+    var count: Long = 0
+        private set
+
+    private var mark: Long = -1
+
+    override fun read(): Int =
+        super.read()
+            .also {
+                if (it != -1) {
+                    count++
+                }
+            }
+
+    override fun read(b: ByteArray?, off: Int, len: Int): Int =
+        super.read(b, off, len)
+            .also { readLen ->
+                if (readLen != -1) {
+                    count += readLen.toLong()
+                }
+            }
+
+    override fun skip(n: Long): Long =
+        super.skip(n)
+            .also { count += it }
+
+    override fun mark(readlimit: Int) {
+        super.mark(readlimit)
+        mark = count
+    }
+
+    override fun reset() {
+        if (!`in`.markSupported()) {
+            throw IOException("Mark not supported")
+        }
+        if (mark == -1L) {
+            throw IOException("Mark not set")
+        }
+
+        super.reset()
+        count = mark.coerceAtLeast(0)
+    }
+
+    suspend fun readRange(range: LongRange): ByteArray {
+        @Suppress("NAME_SHADOWING")
+        val range = range
+            .coerceFirstNonNegative()
+            .requireLengthFitInt()
+
+        if (range.isEmpty())
+            return ByteArray(0)
+
+        return withContext(Dispatchers.IO) {
+            skip(range.first - count)
+            val length = range.last - range.first + 1
+            read(length)
+        }
+    }
+
+}

--- a/r2-shared/src/main/java/org/readium/r2/shared/util/mediatype/SnifferContext.kt
+++ b/r2-shared/src/main/java/org/readium/r2/shared/util/mediatype/SnifferContext.kt
@@ -191,7 +191,12 @@ class SnifferContext internal constructor(
         val archive = contentAsArchive() ?: return null
 
         return withContext(Dispatchers.IO) {
-            tryOrNull { archive.entry(path).read() }
+            tryOrNull {
+                val entry = archive.entry(path)
+                val bytes = entry.read()
+                entry.close()
+                bytes
+            }
         }
     }
 

